rfcs/0007-consistent-vehicle-ids.md

```diff
@@ -0,0 +1,186 @@
+# Summary
+[summary]: #summary
+
+Internal MBTA application publish vehicle IDs with a consistent prefix so as to avoid overlaps.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the MBTA runs vehicles on 5 different modes:
+
+- Bus
+- Commuter Rail
+- Ferry
+- Heavy Rail (Red, Orange, and Blue Lines)
+- Light Rail (Green Line and Mattapan)
+
+Within and between these modes, the identifiers of the vehicle (printed on
+the outside) can overlap, causing confusion in our real-time data. By
+creating consistent prefixes for each mode, we can ensure that the
+identifiers used in our real-time data are unique across all modes.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+<!--
+Explain the proposal as if it was already implemented and you were teaching it to a new developer
+that just joined the team. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how programmers should *think* about the feature, and how it should impact the way they
+  work on this project. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to senior developers and to junior
+  developers. -->
+
+Our vehicle information is published in our GTFS-RT VehiclePositions file. There are three fields which currently contain some vehicle information:
+
+| Field | Description | Values |
+| --- | --- | --- |
+| `id` | Internal system of identification for the vehicle. Should be unique to the vehicle | Heavy rail: the line the car is on plus a hex code that is unique for the consist<br>Light rail: "G-" plus a decimal code that is unique for the consist<br>Commuter Rail: the number of the control coach<br>Bus: the letter "y" plus the number of the bus |
+| `label` | a user visible label - for example the name of a train | Heavy rail: the number of the lead car<br>Light rail: the numbers of each car in the consist, separated by `-`<br>Commuter Rail: the number of the control coach<br>Bus: the number of the bus |
+| `consist` | Information about a single car in a multi-car train | Heavy/rail: `label` contains the number of each car making up the train, in order from front to back |
+
+In particular, bus labels and commuter rail labels overlap, leading us to
+prefix the bus IDs with a `y` to distinguish them. This works as long as we
+only have one type of bus information. However, if we add additional sources
+of bus data (such as contracted shuttle buses) we need another prefix in
+order to distinguish the new vehicles.
+
+To avoid this, this RFC introduces consistent prefixes to give each group of vehicles a unique namespace. The pattern is
+`<vendor>-<mode/line>-<unique identifier>`.
+
+| Mode / Line | Prefix | Old => New |
+| --- | --- | --- |
+| Red Line | `mbta-red` | `R-546C3C33` => `mbta-red-546C3C33` |
```

Comment:
I'm a *little* leery with the idea of putting extraneous information in the IDs, since that means clients might end up relying on that, and changing the ID format in the future would become a breaking change (for those clients, if not by our definition).

For example, the following IDs changed recently which caused problems for various clients:
* `place-dudly` to `place-nubn`
* CR stop IDs to mile markers, e.g. `Lynn` to `place-ER-0115`
* CR trip IDs from e.g. `CR-Saturday-Fall-19-1501` to `CR-483656-506`

But I'm not sure I have a better idea. This is a pretty straightforward way for all the apps to avoid collisions, and I wouldn't want to hash it or anything like that. Maybe a line in the documentation "don't rely on the format, it may change in the future and we don't consider that a breaking change" is all that's needed.

Reply:
[We already list](https://www.mbta.com/developers/v3-api/versioning) changes to the format or value of IDs as something we don't consider a breaking change (though this is specifically a V3 API doc). I don't think this is a totally accurate thing to say, as in the past we've definitely interpreted changes to e.g. stop IDs as a breaking change, and added backwards compatibility for them. IMO we should list each type of resource somewhere and, for each one, indicate whether changes to it are considered backwards-compatible.

Reply:
I think it's also worth teasing apart two separate, but related issues, when thinking about this:
1. Clients assuming that the ID for something like a stop will remain the ID for that stop in the future.
2. Clients assuming that there is meaningful information that they can pull out of the ID rather than simply treating it as an atomic value.

I think 1 is much less of an issue for vehicles than it likely is for, say, stops and trips - clients may care about a certain stop over a long period of time, or a certain trip that happens every day, etc., but not many API / GTFS-rt consumers care about monitoring vehicles in the long term. With respect to 2, I think it's a bit of a mixed bag - I could imagine there maybe being a handful of clients that assume they can (say) remove the "y" from the beginning of a bus vehicle ID to get the number, but on rapid transit there's already nothing very meaningful about the ID aside from the R/G/B/O prefix.

Reply:
Ah, great, thanks for teasing those apart. It helped me clarify my thoughts. Consider my post here then, first and foremost, the obligatory "but should the ID be anything other than a semantically-meaningless opaque blob"?

One of the reasons meaningful names is bad is because the meaning could change (`dudly`) and then the ID needs to change, but as you point out, that's not really a problem here.

But the other reason meaningful names is bad is because it can encourage people to extract meaning from it, when there are better places. My `CR-Saturday-Fall-19-1501` example is relevant here, because when we changed that format we broke an internal Keolis app that relied on the ID having the rating in it.

As you point out, right now people probably aren't using the ID semantically (heh, the Swiftly livemap notwithstanding!), but when we put more info in it, we might encourage them to? I'm just picturing someone displaying vehicles, seeing `mbta-red-ABCDEF` and deciding to group vehicles by `red`, `blue`, etc.

But as Cora points out, we already say that's not something we promise to maintain, so maybe we just 💧 ✋ of the situation and if someone uses it incorrectly it's on them if we decided to change it.

Reply:
Adding on to that: people already can parse out the route out of a heavy rail id, but it feels wrong, like you're not supposed to do it and you're cheating a bit, and you're depending on clearly undocumented behavior. Having human-readable ids makes it feel more trustworthy and documented, even if it isn't intended to be.

One way to have non-overlapping ids without encouraging use would be to have semantic but not-very-readable ids. The one letter prefixes we have right now are great at that, even if they're not a consistent format between the modes.

I'd propose that we change commuter rail ids to be something like `rr1809` (for "regional rail" or "rail road") so they're namespaced, and avoid having to change other already-namespaced modes and without giving the impression that it's meant to be parsed.

Reply:
Hot take: all vehicle IDs are UUIDv5, the operator name is the namespace and the original vehicle ID is the name. Try getting any meaningful information out of that 😄 (yes, I know it's often useful for _us_ specifically to be able to look at a plain vehicle ID in log output or something and see that it's a Red Line car)